src/graphql/schema/User.ts

```diff
@@ -41,11 +49,82 @@ export const UserQuery = extendType({
 			type: "User",
 		});
 		t.list.field("users", {
+			authorize: (_root, _args, ctx) => {
+				const map = ctx.req.session.user!.permissions!.map(
+					(permOnUsers) => permOnUsers.permission.name,
+				);
+				// only if a user can mutate all users he should be able
+				// to query all users
+				return map.includes("canMutateUsers");
+			},
 			resolve: (_root, _args, ctx) => ctx.prisma.user.findMany(),
 			type: "User",
 		});
 	},
 	type: "Query",
 });
 
-// TODO implement auth
+export const LoginMutation = mutationField("login", {
+	args: {
+		password: nonNull(stringArg()),
+		username: nonNull(stringArg()),
+	},
+	resolve: async (_root, { username, password }, ctx) => {
+		const user = await ctx.prisma.user.findFirst({
+			include: {
+				pages: {
+					include: {
+						page: true,
+					},
+				},
+				permissions: {
+					include: {
+						assignedBy: true,
+						permission: true,
+					},
+				},
+				permissionsAssigned: {
+					include: {
+						permission: true,
+						user: true,
+					},
+				},
+			},
+			where: {
+				short: username,
+			},
+		});
+
+		if (!user) {
+			return null;
+		}
+		// compare hash and password
+		if (!(await argon2.verify(user.password, password))) {
+			return null;
```

Comment:
Nexus leaves error handling in resolvers up to the server, and Apollo Server allows native JavaScript error objects to be thrown and will convert that into a valid GraphQL error ([docs](https://www.apollographql.com/docs/apollo-server/data/errors/#throwing-errors)).

Reply:
So you'd suggest throwing a UserInputError instead of returning null?